File: java-workshop/part-2-light-mocks/src/test/java/service/OrderCancellationServiceTest.java
package service;

import domain.CancellationResult;
import domain.Order;
import domain.OrderItem;
import domain.OrderStatus;
import domain.StockReservation;
import org.junit.jupiter.api.Test;
import repository.OrderRepository;

import java.util.List;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.verifyNoInteractions;
import static org.mockito.Mockito.when;

class OrderCancellationServiceTest {

    @Test
    void cancelOrder_refundsPayment_whenOrderExists() {
        OrderRepository orderRepository = mock(OrderRepository.class);
        PaymentService paymentService = mock(PaymentService.class);
        StockService stockService = mock(StockService.class);
        ShippingService shippingService = mock(ShippingService.class);
        OrderCancellationService service = new OrderCancellationService(orderRepository, paymentService, stockService, shippingService);

        Order order = new Order(
            "ORDER-001",
            "CUST-001",
            List.of(new OrderItem("PROD-001", 2, 25.0)),
            OrderStatus.CONFIRMED,
            50.0,
            "TXN-123",
            null
        );

        when(orderRepository.findById("ORDER-001")).thenReturn(order);
        when(paymentService.refundPayment("TXN-123")).thenReturn(true);
        when(shippingService.cancelShipment("TRACK-456")).thenReturn(true);

        CancellationResult result = service.cancelOrder("ORDER-001");

        assertEquals(new CancellationResult.Success(), result);
        verify(paymentService).refundPayment("TXN-123");
        verify(stockService).releaseStock(List.of(new StockReservation("PROD-001", 2, true)));
        verify(orderRepository).updateStatus("ORDER-001", OrderStatus.CANCELLED);
        verifyNoInteractions(shippingService);
    }

    @Test
    void cancelOrder_cancelsShipment_whenTrackingNumberExists() {
        OrderRepository orderRepository = mock(OrderRepository.class);
        PaymentService paymentService = mock(PaymentService.class);
        StockService stockService = mock(StockService.class);
        ShippingService shippingService = mock(ShippingService.class);
        OrderCancellationService service = new OrderCancellationService(orderRepository, paymentService, stockService, shippingService);

        Order order = new Order(
            "ORDER-001",
            "CUST-001",
            List.of(new OrderItem("PROD-001", 2, 25.0)),
            OrderStatus.CONFIRMED,
            50.0,
            "TXN-123",
            "TRACK-456"
        );

        when(orderRepository.findById("ORDER-001")).thenReturn(order);
        when(paymentService.refundPayment("TXN-123")).thenReturn(true);
        when(shippingService.cancelShipment("TRACK-456")).thenReturn(true);

        CancellationResult result = service.cancelOrder("ORDER-001");

        assertEquals(new CancellationResult.Success(), result);
        verify(paymentService).refundPayment("TXN-123");
        verify(stockService).releaseStock(List.of(new StockReservation("PROD-001", 2, true)));
        verify(shippingService).cancelShipment("TRACK-456");
        verify(orderRepository).updateStatus("ORDER-001", OrderStatus.CANCELLED);
    }

    @Test
    void cancelOrder_fails_whenRefundFails() {
        OrderRepository orderRepository = mock(OrderRepository.class);
        PaymentService paymentService = mock(PaymentService.class);
        StockService stockService = mock(StockService.class);
        ShippingService shippingService = mock(ShippingService.class);
        OrderCancellationService service = new OrderCancellationService(orderRepository, paymentService, stockService, shippingService);

        Order order = new Order(
            "ORDER-001",
            "CUST-001",
            List.of(new OrderItem("PROD-001", 2, 25.0)),
            OrderStatus.CONFIRMED,
            50.0,
            "TXN-123",
            "TRACK-456"
        );

        when(orderRepository.findById("ORDER-001")).thenReturn(order);
        when(paymentService.refundPayment("TXN-123")).thenReturn(false);
        when(shippingService.cancelShipment("TRACK-456")).thenReturn(true);

        CancellationResult result = service.cancelOrder("ORDER-001");

        assertEquals(new CancellationResult.Failure("Refund failed"), result);
        verify(paymentService).refundPayment("TXN-123");
        verify(stockService, never()).releaseStock(any());
        verify(shippingService, never()).cancelShipment(anyString());
        verify(orderRepository, never()).updateStatus(anyString(), any());
    }

    @Test
    void cancelOrder_fails_whenOrderDoesNotExist() {
        OrderRepository orderRepository = mock(OrderRepository.class);
        PaymentService paymentService = mock(PaymentService.class);
        StockService stockService = mock(StockService.class);
        ShippingService shippingService = mock(ShippingService.class);
        OrderCancellationService service = new OrderCancellationService(orderRepository, paymentService, stockService, shippingService);

        Order order = new Order(
            "ORDER-001",
            "CUST-001",
            List.of(new OrderItem("PROD-001", 2, 25.0)),
            OrderStatus.CONFIRMED,
            50.0,
            "TXN-123",
            "TRACK-456"
        );

        when(orderRepository.findById("ORDER-001")).thenReturn(null);
        when(paymentService.refundPayment("TXN-123")).thenReturn(true);
        when(shippingService.cancelShipment("TRACK-456")).thenReturn(true);

        CancellationResult result = service.cancelOrder("ORDER-001");

        assertEquals(new CancellationResult.Failure("Order not found"), result);
        verify(orderRepository).findById("ORDER-001");
        verifyNoInteractions(paymentService, stockService, shippingService);
    }
}

File: kotlin-workshop/part-3-mocks-clean/src/test/kotlin/service/OrderCancellationServiceTest.kt
package service

import domain.*
import io.mockk.*
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import repository.OrderRepository
import kotlin.test.assertEquals

class OrderCancellationServiceTest {

    private val orderRepository = mockk<OrderRepository>()
    private val paymentService = mockk<PaymentService>()
    private val stockService = mockk<StockService>()
    private val shippingService = mockk<ShippingService>()

    private val service = OrderCancellationService(
        orderRepository,
        paymentService,
        stockService,
        shippingService
    )

    @BeforeEach
    fun setup() {
        clearMocks(orderRepository, paymentService, stockService, shippingService)
    }

    @Test
    fun `cancel order before shipping - refunds and releases stock`() {
        // Given
        val order = Order(
            id = "ORDER-001",
            customerId = "CUST-001",
            items = listOf(OrderItem("PROD-001", 2, 25.0)),
            status = OrderStatus.CONFIRMED,
            transactionId = "TXN-123",
            trackingNumber = null
        )

        every { orderRepository.findById("ORDER-001") } returns order
        every { stockService.releaseStock(any()) } just Runs
        every { paymentService.refundPayment("TXN-123") } returns true
        every { orderRepository.updateStatus("ORDER-001", OrderStatus.CANCELLED) } just Runs

        // When
        val result = service.cancelOrder("ORDER-001")

        // Then
        assertEquals(CancellationResult.Success, result)

        verifySequence {
            orderRepository.findById("ORDER-001")
            stockService.releaseStock(any())
            paymentService.refundPayment("TXN-123")
            orderRepository.updateStatus("ORDER-001", OrderStatus.CANCELLED)
        }
        confirmVerified(shippingService)
    }

    @Test
    fun `cancel order with shipment in progress - cancels everything`() {
        // Given
        val order = Order(
            id = "ORDER-002",
            customerId = "CUST-001",
            items = listOf(
                OrderItem("PROD-001", 1, 50.0),
                OrderItem("PROD-002", 3, 15.0)
            ),
            status = OrderStatus.CONFIRMED,
            transactionId = "TXN-456",
            trackingNumber = "TRACK-789"
        )

        every { orderRepository.findById("ORDER-002") } returns order
        every { stockService.releaseStock(any()) } just Runs
        every { paymentService.refundPayment("TXN-456") } returns true
        every { shippingService.cancelShipment("TRACK-789") } returns true
        every { orderRepository.updateStatus("ORDER-002", OrderStatus.CANCELLED) } just Runs

        // When
        val result = service.cancelOrder("ORDER-002")

        // Then
        assertEquals(CancellationResult.Success, result)

        verifySequence {
            orderRepository.findById("ORDER-002")
            stockService.releaseStock(any())
            paymentService.refundPayment("TXN-456")
            shippingService.cancelShipment("TRACK-789")
            orderRepository.updateStatus("ORDER-002", OrderStatus.CANCELLED)
        }
    }

    @Test
    fun `cannot cancel shipped order`() {
        // Given
        val order = Order(
            id = "ORDER-003",
            customerId = "CUST-001",
            items = listOf(OrderItem("PROD-001", 1, 100.0)),
            status = OrderStatus.SHIPPED,
            transactionId = "TXN-789"
        )

        every { orderRepository.findById("ORDER-003") } returns order

        // When
        val result = service.cancelOrder("ORDER-003")

        // Then
        assertEquals(CancellationResult.Failure("Cannot cancel shipped order"), result)
        verify(exactly = 0) { paymentService.refundPayment(any()) }
        verify(exactly = 0) { stockService.releaseStock(any()) }
    }

    @Test
    fun `order not found returns failure`() {
        // Given
        every { orderRepository.findById("UNKNOWN") } returns null

        // When
        val result = service.cancelOrder("UNKNOWN")

        // Then
        assertEquals(CancellationResult.Failure("Order not found"), result)
    }
}

